File: tests/client_compatibility/dotnet/BasicConnectionTest.cs
using Npgsql;
using Xunit;

namespace PGWireCompatibility;

/// <summary>
/// Test basic Npgsql connection to IRIS PGWire server.
///
/// Tests P0 Handshake Protocol:
/// - SSL negotiation
/// - StartupMessage processing
/// - Authentication
/// - ReadyForQuery state
/// </summary>
public class BasicConnectionTest : IDisposable
{
    private readonly string connectionString;

    public BasicConnectionTest()
    {
        var host = Environment.GetEnvironmentVariable("PGWIRE_HOST") ?? "localhost";
        var port = Environment.GetEnvironmentVariable("PGWIRE_PORT") ?? "5432";
        var database = Environment.GetEnvironmentVariable("PGWIRE_DATABASE") ?? "USER";
        var username = Environment.GetEnvironmentVariable("PGWIRE_USERNAME") ?? "test_user";
        var password = Environment.GetEnvironmentVariable("PGWIRE_PASSWORD") ?? "test";

        connectionString = $"Host={host};Port={port};Database={database};Username={username};Password={password}";
    }

    public void Dispose()
    {
        // Cleanup if needed
    }

    [Fact]
    public async Task TestBasicConnection()
    {
        // GIVEN: PGWire server is running
        // WHEN: Attempting to connect
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // THEN: Connection should be established
        Assert.Equal(System.Data.ConnectionState.Open, conn.State);
        Assert.NotNull(conn.DataSource);
        Assert.NotNull(conn.Database);
    }

    [Fact]
    public async Task TestConnectionWithConnectionStringBuilder()
    {
        // GIVEN: Connection configured via builder
        var builder = new NpgsqlConnectionStringBuilder
        {
            Host = Environment.GetEnvironmentVariable("PGWIRE_HOST") ?? "localhost",
            Port = int.Parse(Environment.GetEnvironmentVariable("PGWIRE_PORT") ?? "5432"),
            Database = Environment.GetEnvironmentVariable("PGWIRE_DATABASE") ?? "USER",
            Username = Environment.GetEnvironmentVariable("PGWIRE_USERNAME") ?? "test_user",
            Password = Environment.GetEnvironmentVariable("PGWIRE_PASSWORD") ?? "test",
            SslMode = SslMode.Disable // Plain text for testing
        };

        // WHEN: Connecting with builder
        await using var conn = new NpgsqlConnection(builder.ConnectionString);
        await conn.OpenAsync();

        // THEN: Connection succeeds
        Assert.Equal(System.Data.ConnectionState.Open, conn.State);
    }

    [Fact]
    public async Task TestConnectionPooling()
    {
        // GIVEN: Connection string with pooling enabled
        var builder = new NpgsqlConnectionStringBuilder(connectionString)
        {
            Pooling = true,
            MinPoolSize = 1,
            MaxPoolSize = 10
        };

        // WHEN: Opening multiple connections
        await using var conn1 = new NpgsqlConnection(builder.ConnectionString);
        await using var conn2 = new NpgsqlConnection(builder.ConnectionString);

        await conn1.OpenAsync();
        await conn2.OpenAsync();

        // THEN: Both connections should be open
        Assert.Equal(System.Data.ConnectionState.Open, conn1.State);
        Assert.Equal(System.Data.ConnectionState.Open, conn2.State);
    }

    [Fact]
    public async Task TestServerVersion()
    {
        // GIVEN: Active connection
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // WHEN: Querying server version
        var serverVersion = conn.ServerVersion;

        // THEN: Version should be available
        Assert.NotNull(serverVersion);
        Assert.NotEmpty(serverVersion);
        Console.WriteLine($"Server Version: {serverVersion}");
    }

    [Fact]
    public async Task TestConnectionInfo()
    {
        // GIVEN: Active connection
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // WHEN: Querying connection info
        // THEN: Connection properties should be accessible
        Assert.NotNull(conn.Host);
        Assert.NotNull(conn.Database);
        Assert.True(conn.Port > 0);

        Console.WriteLine($"Host: {conn.Host}");
        Console.WriteLine($"Port: {conn.Port}");
        Console.WriteLine($"Database: {conn.Database}");
        Console.WriteLine($"Server Version: {conn.ServerVersion}");
    }

    [Fact]
    public async Task TestMultipleSequentialConnections()
    {
        // GIVEN: Connection string
        // WHEN: Opening and closing connections sequentially
        for (int i = 0; i < 3; i++)
        {
            await using var conn = new NpgsqlConnection(connectionString);
            await conn.OpenAsync();

            // THEN: Each connection should succeed
            Assert.Equal(System.Data.ConnectionState.Open, conn.State);
            await conn.CloseAsync();
            Assert.Equal(System.Data.ConnectionState.Closed, conn.State);
        }
    }
}

File: tests/client_compatibility/jdbc/src/test/java/com/intersystems/iris/pgwire/BasicConnectionTest.java
package com.intersystems.iris.pgwire;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.AfterAll;
import static org.junit.jupiter.api.Assertions.*;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;

/**
 * Test basic PostgreSQL JDBC connection to IRIS PGWire server.
 *
 * Tests P0 Handshake Protocol:
 * - SSL negotiation
 * - StartupMessage processing
 * - Authentication
 * - ReadyForQuery state
 */
public class BasicConnectionTest {

    private static String host;
    private static int port;
    private static String database;
    private static String username;
    private static String password;

    @BeforeAll
    public static void setup() {
        host = System.getProperty("pgwire.host", "localhost");
        port = Integer.parseInt(System.getProperty("pgwire.port", "5432"));
        database = System.getProperty("pgwire.database", "USER");
        username = System.getProperty("pgwire.username", "test_user");
        password = System.getProperty("pgwire.password", "test");
    }

    @Test
    public void testBasicConnection() throws SQLException {
        // GIVEN: PGWire server is running
        String url = String.format("jdbc:postgresql://%s:%d/%s", host, port, database);

        // WHEN: Attempting to connect
        try (Connection conn = DriverManager.getConnection(url, username, password)) {
            // THEN: Connection should be established
            assertNotNull(conn, "Connection should not be null");
            assertFalse(conn.isClosed(), "Connection should be open");
            assertTrue(conn.isValid(5), "Connection should be valid within 5 seconds");
        }
    }

    @Test
    public void testConnectionWithProperties() throws SQLException {
        // GIVEN: Connection properties configured
        String url = String.format("jdbc:postgresql://%s:%d/%s", host, port, database);
        Properties props = new Properties();
        props.setProperty("user", username);
        props.setProperty("password", password);
        props.setProperty("ApplicationName", "JDBC-Compatibility-Test");
        props.setProperty("ssl", "false"); // Plain text for testing

        // WHEN: Connecting with properties
        try (Connection conn = DriverManager.getConnection(url, props)) {
            // THEN: Connection succeeds with properties
            assertNotNull(conn);
            assertFalse(conn.isClosed());
        }
    }

    @Test
    public void testConnectionMetadata() throws SQLException {
        // GIVEN: Active connection
        String url = String.format("jdbc:postgresql://%s:%d/%s", host, port, database);

        try (Connection conn = DriverManager.getConnection(url, username, password)) {
            // WHEN: Querying connection metadata
            var metadata = conn.getMetaData();

            // THEN: Metadata should be available
            assertNotNull(metadata, "Database metadata should not be null");
            assertTrue(metadata.getDatabaseProductName().contains("IRIS") ||
                      metadata.getDatabaseProductName().contains("PostgreSQL"),
                      "Database product should be IRIS or PostgreSQL-compatible");

            System.out.println("Database Product: " + metadata.getDatabaseProductName());
            System.out.println("Database Version: " + metadata.getDatabaseProductVersion());
            System.out.println("Driver Name: " + metadata.getDriverName());
            System.out.println("Driver Version: " + metadata.getDriverVersion());
        }
    }

    @Test
    public void testAutoCommitDefault() throws SQLException {
        // GIVEN: New connection
        String url = String.format("jdbc:postgresql://%s:%d/%s", host, port, database);

        try (Connection conn = DriverManager.getConnection(url, username, password)) {
            // THEN: Auto-commit should be enabled by default
            assertTrue(conn.getAutoCommit(), "Auto-commit should be enabled by default");
        }
    }

    @Test
    public void testReadOnlyMode() throws SQLException {
        // GIVEN: Active connection
        String url = String.format("jdbc:postgresql://%s:%d/%s", host, port, database);

        try (Connection conn = DriverManager.getConnection(url, username, password)) {
            // WHEN: Setting read-only mode
            conn.setReadOnly(false);

            // THEN: Read-only status should be queryable
            assertFalse(conn.isReadOnly(), "Connection should not be read-only");
        }
    }
}
